code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/RegionMonitor.java

```diff
@@ -0,0 +1,122 @@
+        for(Region region : regions) {
+            regionMap.put(region.getId(), region);
+            if(region.getLocal()) {
+                localRegion = region;
+            }
+        }
+        List<AccountLink> accountLinks = objectManager.find(AccountLink.class, ACCOUNT_LINK.REMOVED, null, ACCOUNT_LINK.LINKED_REGION_ID, 
+            new Condition(ConditionType.NOTNULL));
+        Map<String, ExternalRegion> externalRegionMap = new HashMap<String, ExternalRegion>();
+        Map<String, ExternalProject> externalProjectMap = new HashMap<String, ExternalProject>();
+        for(AccountLink link : accountLinks) {
+            if (invalidStates.contains(link.getState())) {
+                continue;
+            }
+            // targetRegion not present
+            Region targetRegion = regionMap.get(link.getLinkedRegionId());
+            if(targetRegion == null) {
+                delete(link);
+                continue;
+            }
+            try {
+                // localRegion in targetRegion not present
+                ExternalRegion externalRegion = null;
+                String externalRegionKey = String.format("%s$%s", targetRegion, localRegion.getName());
+                if(externalRegionMap.containsKey(externalRegionKey)) {
+                    externalRegion = externalRegionMap.get(externalRegionKey);
+                } else {
+                    externalRegion = RegionUtil.getExternalRegion(targetRegion, localRegion.getName(), jsonMapper);
+                    externalRegionMap.put(externalRegionKey, externalRegion); 
+                }
+                if(externalRegion == null) {
+                    delete(link);
+                    continue;
+                }
+                // environment not present or its uuid changed 
+                ExternalProject externalProject = null;
+                String externalProjectKey = String.format("%s$%s", targetRegion, link.getLinkedAccount());
+                if(externalProjectMap.containsKey(externalProjectKey)) {
+                    externalProject = externalProjectMap.get(externalProjectKey);
+                } else {
+                    externalProject = RegionUtil.getTargetProjectByName(targetRegion, link.getLinkedAccount(), jsonMapper);
+                    externalProjectMap.put(externalProjectKey, externalProject);
+                }
+                String storedUUID = DataAccessor.fieldString(link, "uuid");
+                if(externalProject == null || (storedUUID!=null && !externalProject.getUuid().equals(storedUUID))) {
+                    delete(link);
+                }
+            } catch (Exception ex) {
+                log.info("Failed to monitor account link for %s - %s", link.getId(), ex);
+            }
+        }
+        List<Account> accounts = objectManager.find(Account.class, ACCOUNT.REMOVED, new Condition(ConditionType.NULL));
+        for(Account account : accounts) {
+            regionService.reconcileExternalLinks(account.getId());
```

Comment:
lets wrap it in try/catch too. There can be process cancel exception thrown underneath